content/posts/faster-rfft-python/index.md

```diff
@@ -0,0 +1,386 @@
+
+As shown in the performance plot above, our first attempt to optimize the RFFT computation resulted in about 2-12% speed-up for most input sizes. It is not clear to me why the performance drops for inputs with sizes `2^13` to `2^15`. I am going to skip the investigation for those cases. Let's see if we can do better!!
+
+
+### Second Attempt: Tell RFFT to avoid copy (v2)
+According to [pyfftw.builders.rfft documentation](https://pyfftw.readthedocs.io/en/latest/source/pyfftw/builders/builders.html#module-pyfftw.builders): 
+
+> By default, these functions will always create a copy (and sometimes more than one) of the passed-in input array. This is because the creation of the pyfftw.FFTW object generally destroys the contents of the input array.
+
+> This only influences a copy during the creation of the object
+
+In our case, the byte-aligned array `real_arr` is used when creating the RFFT object. And, then, it is filled with the contents of `T` right before executing the RFFT object. Therefore, at the time of creating the RFFT object, the contents of `real_arr` are not important as they will be overwritten. We do this by setting the `avoid_copy` parameter to `True` to avoid unnecessary copying of data at the time of creating the RFFT object.
+
+```python
+class rfft_caller_v2:
+    def __init__(self):
+        self.real_arr = None
+
+    def __call__(self, T):
+        if self.real_arr is None or len(T) != len(self.real_arr):
+            self.real_arr = pyfftw.empty_aligned(len(T), dtype='float64')
+        
+        rfft_obj = pyfftw.builders.rfft(self.real_arr, avoid_copy=True)
+        self.real_arr[:] = T
+        rfft_obj.execute()
+        
+        return rfft_obj.output_array
+
+_rfft_caller_v2 = rfft_caller_v2()
+```
+
+Let's update the dictionary `rfft_callers` to include this new version, and check out the performance improvements.
+
+```python
+rfft_callers = {
+    'v0': rfft_caller_v0,
+    'v1': rfft_caller_v1,
+    'v2': rfft_caller_v2,
+}
+```
+
+In the performance plot below, the orange line represents the performance of this new version (v2). The higher the line, the better the performance.
+
+![Performance Gain](Figure_v2.png)
+
+As shown in the performance plot above, our second attempt to optimize the RFFT computation resulted in about 5%-15% speed-up relative to the baseline. 
+
+
+### Third Attempt: Use pyfftw.FFTW directly (v3)
+In the previous versions, we used `pyfftw.builders.rfft` to create the RFFT object. As mentioned in [pyfftw.builders documentation](https://pyfftw.readthedocs.io/en/latest/source/pyfftw/builders/builders.html), the goal of builder functions is to provide a convenient interface for creating FFTW objects. Under the hood, any builder function returns an object that is an instance of `_FFTWWrappe`, which is a child class of `pyfftw.FFTW`. One potential optimization is to use `pyfftw.FFTW` directly. This might save some overhead associated with the builder functions.
```

Comment:
>  any builder function returns an object that is an instance of `_FFTWWrappe`, which is a child class of `pyfftw.FFTW`.

" any builder function returns an object that is an instance of `pyfftw.FFTW`"

See: https://pyfftw.readthedocs.io/en/latest/source/pyfftw/builders/builders.html#module-pyfftw.builders